Fixing errors in canopy calculation (#503)
ImperialCollegeLondon/pyrealm: docs/source/users/demography/canopy.md

from pyrealm.demography.tmodel import StemAllometry
```

The `canopy` module in `pyrealm` is used to calculate a vertically structured model of
leaf distribution across all of cohorts within a [plant community](./community.md),
where a cohort is defined as:

* a number of individuals,
* with identical diameter at breast height (DBH, $D$),
* of the same [plant functional type (PFT)](./flora.md),
* that share the same [stem allometry](./t_model.md) and [crown model](./crown.md).

The purpose of the `canopy` module is two-fold:

1. to calculate the vertical distribution of crown and leaf area, including the
   partitioning of canopy into discrete vertical layers, and
2. to estimate light capture through canopy layers.

This page describes the first of the roles and [light capture](./light_capture.md)
is presented separately.

## Vertically structured canopies

> The simplest "big leaf" model of a canopy, represents the entire area of the crown
> ($A_c$) as a single flat disc at the top of stem. However, the canopy model uses the

Reviewer: ```suggestion
Under the simplest "big leaf" model of a canopy, the area of an individual crown ($A_c$)
is represented as a single flat disc at the top of the stem. However, the canopy model uses the
```

> ($A_c$) as a single flat disc at the top of stem. However, the canopy model uses the

Reviewer: ```suggestion
($A_c$) as a single flat disc at the top of the stem. However, the canopy model uses the
```

crown shape described in the [crown model](./crown.md) to model how leaf area
accumulates vertically through the cohorts of a community.

The canopy model can profile community crown area at given heights - useful to generate
detailed plots of vertical structure - but also implements a specific model that
partitions cohort crown area vertically into a series of closed canopy layers. The
partitioning follows the perfect plasticity approximation (PPA) model
{cite}`purves:2008a`, which assumes that plants are always able to plastically arrange
their crown within the broader canopy of the community to maximise their crown size and

> fill the available space. The available space $A$ is simply the plot size that the

Reviewer: These sentences imply that PPA only works horizontally, but a crown will grow both horizontally (increase the projected crown area) and vertically (increase the crown length). 

Edit: After going through the rest of the documentation below, would it be correct to say that the model first maximises horizontal expansion, and if there's extra crown area that did not fit inside the "total area" then this extra gets added to the layer above (representing vertical expansion)? 
Related to this, I wonder if ecologically speaking this "crown-formation process" would take place bottom-up (i.e., upward crown growth is stimulated) or top-down (i.e., growth is suppressed at the bottom of the crown due to lack of space/competition, in other words the extra does not get added but gets removed instead). Not suggesting any changes here, just interesting to think about.

Author: Good point - will extend the description.

That's an interesting point about the growth of canopies - the stem has a fixed vertical shape for the crown area but as the stem grows of course the amount of crown at different points changes. The model _only_ follows that shape - trees do not grow into available space unless they have crown area within a particular layer. I guess the end result is that the shape represents ecological constraints on where a particular growth form can place leaves. That growth form is less plastic than reality.

Reviewer: Thanks for clarifying, I now also realise that the first part of my question was not relevant, so no more questions here.

community is growing in - all crown area is constrained to occur within the plot.

> If the total canopy area of all individuals within cohorts is less than $A$ then the

Reviewer: Should it not be "across cohorts" instead of "within cohorts"? If this calculation is done for each cohort separately (so within cohorts) then wouldn't this result in excessively large crown areas?

Edit: I think I understand why you used "within cohorts", it seems I misunderstood this part and thought you were describing here how the total projected crown area within a cohort is calculated (which is actually derived from the crown module). To avoid others making this same mistake, perhaps you can reiterate that this total crown area is derived from the crown module to smoothen this transition.

Reviewer: This question is no longer relevant.

canopy model is complete - in this case, the predications are effectively the same as

> the big-leaf model. However if the total crown area across the stems is greater than $A$

Reviewer: Okay, after realising my edit in the comment above this second part also makes a lot more sense now. I still have a few questions here:
- So if this calculation happens within cohorts, it seems unlikely to me that the projected crown area of a particular cohort at a given height exceeds the total area of the grid. Am I missing something here? Can it be that this step works at community level and that the reference to cohorts is confusing?
- If this calculation/designation of total crown area happens within cohorts, and if this step is then repeated for each cohort, wouldn't that result in the overlapping of a certain available spacy by multiple tree crowns at once (which wouldn't be very realistic)? My guess is that by working with the total grid area you can avoid this (i.e., a certain space within a layer can only be filled once) but I am then curious as to how you decide which cohort gets priority to fill up that space first?

Reviewer: When saying "total crown area across the stems is greater than A" does this refer to the total crown area at a specific height (I'm thinking of it like a slice of the crown at a specific height, done for all individuals in that cohort) or does this refer to the cumulative crown area of all the "slices" in a crown at all the different heights that should contain crown, summed across all individuals in that cohort?

Author: The terminology is horrible in this section 😄 

I think avoiding referring to cohorts is worth doing, except to note that:

* a community is made up of a set of stems and the canopy model works across all stems simultaneously.
* those stems may be grouped into cohorts with identical DBH and PFT, but actually that's kind of irrelevant to the canopy model.

There is no priority system beyond the crown area at height $z$ function. The sum of crown area is across all stems simultaneously, so that allocation to layers purely follows that total accumulating area.

Reviewer: Thanks! This part is clear now.

Author: @arne-exe Thanks a lot for going through and saying where questions have been resolved - super helpful for me to know when I don't need to take any further action. There is also the "resolve conversation" button below, but having the person who asked the question specifically say that it is resolved is clearer, so much appreciated.

> then the model finds heights at which successive layers fill up the available area,

Reviewer: Out of curiosity, what is the depth/thickness of a layer?

Author: It is variable. The first layer starts at the top of the highest stem and closes at the first multiple of plot area. Then each lower layer is between successive lower layers until the last layer closes at ground level.

Reviewer: This makes sense now!

giving closed canopy layers.

> We can do this using the projected crown area ${A}_{p}(z)$ from the crown model, which

Reviewer: You nicely explain it here, perhaps worth moving this paragraph up a bit as it would help to interpret earlier paragraphs.

describes how crown area accumulates with height $z$ from the top of the stem to the
ground, given the crown shape of the tree. If we sum the projected crown area across all

> stems in a community, we generate a prediction of the total community crown area at a

Reviewer: Here it is clear that you work with the summed projected crown area across all stems in the community. If you would move this paragraph up then the previous paragraph becomes a lot easier to interpret (i.e., it is clear that you are working with community projected crown area and not within cohorts).

given height $z$:

$$
\sum_1^{N_s}{ A_p(z)}
$$

We then need to find the heights $z^*_l$ at which the total community crown area

> occupies multiples of the available area $A$, giving the heights at which canopy

Reviewer: The purpose of this step is not entirely clear. Is the goal here to define the heights where the total community crown area is more than 1 time the available area? If so, does this mean that when it is more than 1 time the available area the additional crown area gets added to the next layer? Looking at the equations that follow it doesn't seem like the total community crown area for a given height is "updated" based on any extra from previous layers. It could be that I'm misunderstanding things here.

Author: The projected community crown area accumulates from the top of the highest stem down to ground area. The heights $z^*_l$ are then the points at which the community area exceeds multiples of $A$ or $A(1-f_G)$.

I think starting with that curve might be a good place to go from.

Reviewer: Yup, this makes sense now!

layers close for canopy layers $l = 1, ..., l_m$. This is given as the equality:

$$
\sum_1^{N_s}{ A_p(z^*_l)} = l A
$$

An extra detail is that the canopy model includes a community-level canopy gap
fraction ($f_G$) that captures the overall proportion of the canopy area that is left
unfilled by canopy. This gap fraction, capturing processes such as crown shyness,
describes the proportion of open sky visible from the forest floor. This term reduces

> the amount of $A$ that can be occupied by crown area before a layer closes:

Reviewer: Another thing to consider is that the total available area would also need to be corrected for the area occupied by stems, as leaves cannot grow inside a stem, although this area is likely to be relatively small at plot level.

Author: I think this is effectively absorbed into the leaf area index? 

The projected crown area is not the total amount of leaf area in the layer - it's the area occupied by the crown in the layer. The same (confusingly) is true of the projected leaf area. The actual absorption is a function of the total leaf surface within the layer which is governed by the 2D projected area and the LAI in the Beer-Lambert law. So any single layer is really a volume containing stems and leaves and the total light absorption (which would be modified by stems) is captured by the LAI. A particularly stemmy PFT might have a lower LAI?

Thoughts?

Reviewer: I guess it would depend on how LAI data is collected, and whether or not they corrected for stems (e.g., in hemisperical images). 

The other thought I had on this was whether gaps in the crown/canopy are corrected for twice (as the model uses both LAI and gap fraction)? I'm sure this has crossed your mind already, but I couldn't quite figure out how one correction relates to the other, as they are used in different modules (crown and light capture) but the light capture module relies on the crown leaf area (if I'm not mistaken).

Author: There isn't a double counting - I'll try and clarify this somewhere. Does this text help?

* The LAI is "just" a characteristic value of the PFT setting the amount of leaf area per m2 of crown area.
* The crown gap fraction effectively tunes where vertically in the crown that LAI is deployed. The crown gaps do not push all the way down to the ground - they only allow light to penetrate more deeply into the crown.
* The canopy gap fraction is never intercepted by any crown - all light within the canopy gap fraction passes unimpeded from the top of the canopy to the ground.

This page describes the model of the vertical canopy structure and the [light capture
model](./light_capture.md) is presented separately.

## Vertically structured canopies

The simplest "big leaf" model of a canopy, represents the entire crown crown ($A_c$) of
an individual tree as a single flat disc at the top of the stem. However, in `pyrealm`
the canopy model uses the crown shape described in the [crown model](./crown.md) to
model how leaf area accumulates vertically through a community.

### Community definition

To recap, a [plant community](./community.md) consists of a number of individual stems
that are growing together in a location. The community structure groups individuals
together into cohorts that are defined as:

* a number of individuals,
* with identical diameter at breast height (DBH, $D$),
* of the same [plant functional type (PFT)](./flora.md),
* and hence have the same [stem allometry](./t_model.md) and [crown model](./crown.md).

However, for the purposes of the canopy model, it is much simpler to consider the
community as a collection of individual stems, some of which just happen to share
identical properties.

### Community projected crown area

> Each individual stem in a community has an projected crown area that describes how the
> crown area of that stem accumulates with height ($z$) from the top of the tree to the
> ground. The form of that curve varies with the size and PFT of the stem. If we have a
> community with $N_s$ individuals, then each stem $i = 1, ..., N_s$ has a projected crown
> area function $A_{p}(z)_i$.

Reviewer: ```suggestion
Each individual stem in a community has a crown area profile that describes how the
projected crown area of that stem accumulates with height ($z$) from the top of the tree
to the ground. For a community with $N_s$ individuals, the crown area profile of each stem
$i = 1, ..., N_s$ is captured by the function $A_{p}(z)_i$, where the shape of this curve
varies with the size and PFT of the stem. Summing the projected crown area for all values
of ($z$) contained within an individual's crown yields the total crown area.
```

Reviewer: The way I interpret this is that the crown area function predicts a projected crown area at a specific height (i.e., a particular slice of the crown at that height). This function is then repeated for all heights contained within a tree's stem height, which combined gives the crown area profile (and when summed gives the total projected crown area for that individual).

Reviewer: This is more related to the crown module, but when looking at the figure below containing the community crown area, it seems like crowns can run all the way to the ground. I would assume that each stem has a crown-free section. Is this something that is included in the model (for example by restricting the lower limit of z values depending on stem height)?

Author: Hmmm. This is something I need to clarify in the crown model. If I've got this right, the crown is technically only the upper surface of the overall crown shape (above the maximum crown radius). So the stem below that level, which has a relative height of  1 - the crown fraction, is crown free.

BUT... The crown gap fraction pushes leaf area down below that crown surface layer, so trees where the crown gap fraction is > 0 will have some leaf surface below that crown fraction height.

I might make this a new issue, just to get this closed down!

Reviewer: > Hmmm. This is something I need to clarify in the crown model. If I've got this right, the crown is technically only the upper surface of the overall crown shape (above the maximum crown radius). So the stem below that level, which has a relative height of 1 - the crown fraction, is crown free.
> 

Ah, you're right, I didn't realise this before. Indeed, the figure right above where you introduce the PPA model shows that below the maximum crown radius the cumulative projected crown area does not increase for a particular individual/cohort.

If we take the sum across all individuals of this function at a height $z$, we get the
total community projected crown area across all individual stems at that same height:

$$
C_p(z) = \sum_{j=1}^{N_s}{A_{p}(z)_i}
$$

> To demonstrate this, the code below creates a plant community: there are 12 individual
> stems, grouped into 3 cohorts with different stem sizes and crown shapes.

Reviewer: ```suggestion
To demonstrate this, the code below creates a plant community that consists of 12 individual
stems, grouped into 3 cohorts with different stem sizes and crown shapes.
```

        pft_names=np.array(["short", "short", "tall"]),
    ),
)
```

The total crown area of the individuals in each of the three cohorts is shown below:

```{code-cell} ipython3
community.stem_allometry.crown_area
```

The total crown area across the community is then the sum of those crown areas across
all individuals:

```{code-cell} ipython3
total_crown_area = np.sum(
    community.stem_allometry.crown_area * community.cohorts.n_individuals, keepdims=True
)
total_crown_area
```

> So, the maximum projected community crown area is around 120 m2. The plot below shows

Reviewer: ```suggestion
So, the total community crown area is around 120 m2. The plot below shows
```

    for stem_loc in stem_locations:
        ax1.add_patch(Polygon(crown + np.array([stem_loc, 0]), color="#00550055"))

ax1.autoscale_view()
add_hvlines(community.stem_allometry.stem_height, ax1, False)
ax1.set_ylabel("Height (m)")

ax2.plot(crown_area, hghts)
add_hvlines(community.stem_allometry.stem_height, ax2, False)
add_hvlines([0, total_crown_area], ax2, True)
_ = ax2.set_xlabel("Community projected crown area ($C_{p}(z), m^2$)")

plt.tight_layout()
```

If the community is growing in an area greater than 120 m2, then the individuals can
avoid overlapping any of their crown area. However, this is not possible if the
community is growing in a smaller area, and some crown area must overlie other parts
of the community, creating a vertical structure.

The canopy module implements the **perfect plasticity approximation (PPA)** model
{cite}`purves:2008a` to generate this structure. The PPA model assumes that all the
individuals within the community are able to plastically arrange their crown at each height
within the broader canopy of the community to fill available space. When the available
space ($A$) is filled, crown area lower in the canopy forms another layer until another

> area $A$ is filled and this repeats down to the ground.

Reviewer: I understand this now. I was initially focusing too much on how this would work for individual trees at each specific height (z), but that's not how it works.

Instead, it defines canopy layers by operating across all individuals and across multiple height values. For each height z, the projected crown area is summed across all individuals at that height. If the total does not exceed the available horizontal space, the model proceeds to the next lower height and adds the crown area at that level to a cumulative "bucket." This continues until the cumulative projected crown area exceeds the available space, at which point the current canopy layer is considered full. A new layer (or "bucket") is then started, and the process repeats down through the canopy.

Author: Exactly. The "next lowest height" is not found from a fixed set of heights, but instead a numeric method is used to wiggle the height around until the bucket pretty much exactly fills (to about 1 cm2 precision by default). But otherwise this is a great summary.

> space ($A$) is filled, crown area lower in the canopy forms another layer until another
> area $A$ is filled and this repeats down to the ground.

Reviewer: ```suggestion
horizontal space ($A$) is filled by the cumulative crown area across all individuals at
one or more height levels ($z$), that canopy layer is considered "full" (or "closed").
The remaining crown area below this layer then begins to accumulate into a new layer.
This process repeats downward until the entire canopy structure is formed down to the ground.
```

the amount of $A$ that can be occupied by crown area before a layer closes:

$$
C_p(z^*_l) = l A (1- f_G)
$$

The values of $z^*_l$ cannot be found analytically, so we use numerical methods to find
values of $z^*_l$ for $l = 1,..., l_m$ that satisfy:

$$
C_p(z^*_l) - l A(1 - f_G) = 0
$$

The total number of layers $l_m$ in a canopy, where the final layer may not be fully
closed, can be found given the total crown area across stems as:

$$
l_m = \left\lceil \frac{\sum_1^{N_s}{A_c}}{A(1 - f_G)}\right\rceil
$$

The community above has an available space $A=32\;m^2$. If we fit the PPA model to this
community we get the following heights:

```{code-cell} ipython3
# Fit the canopy model
canopy_ppa = Canopy(community=community, fit_ppa=True)
canopy_ppa.heights
```

We plot those layer heights below. The vertical dashed lines in the right hand plot show
the values $lA$ at which the cumulative space across layers forms closed layers: that is
$32 \times 1 = 32$, $32 \times 2 = 64$, etc. . The horizontal dashed lines then show
those calculated vertical heights. These are the canopy heights at which the vertical

> lines intersect the cumulativecommunity projected crown area: where the total crown area

Reviewer: ```suggestion
lines intersect the cumulative community projected crown area: where the total crown area
```

Reviewer: This figure is very neat. Once it clicks how everything works it's a very nice visualisation.

Author: Good - glad it helps. I think it is a lot clearer now this is set out like this

)

# Create a canopy plot
for idx, crown in enumerate(profiles):

    # Get spaced but slightly randomized stem locations
    n_stems = community.cohorts.n_individuals[idx]
    stem_locations = np.linspace(0, 10, num=n_stems) + np.random.normal(size=n_stems)

    # Plot the crown model for each stem
    for stem_loc in stem_locations:
        ax1.add_patch(Polygon(crown + np.array([stem_loc, 0]), color="#00550055"))

add_hvlines(canopy_ppa.heights, ax1, False)
ax1.set_ylabel("Height (m)")

# Plot community projected crown area
ax2.plot(crown_area, hghts)
ax2.set_xlabel("Cumulative crown area (m2)")

# Add lines to show PPA closure areas and heights
area_values = np.arange(1, 5) * 32
add_hvlines(canopy_ppa.heights, ax2, False)
add_hvlines(area_values, ax2, True)

# Add secondary axes to give values
add_second_axis(canopy_ppa.heights.squeeze(), ax2, "$z^*_{idx} = ${val:0.2f}", False)
add_second_axis(area_values, ax2, "$A_{idx} = ${val}", True)

plt.tight_layout()
```

We can look in detail at the how much crown area is in each layer from each individual.
The canopy model object contains a crown profile which records the projected crown

> area at each height:

Reviewer: ```suggestion
area at each height, summarised at cohort level:
```

Reviewer: Looking at the example array this seems to give a summary at cohort level, not individual

Author: Nope - that is individual level. The bottom row matches the total individual crown area from just the allometry above . I've added a note.

```{code-cell} ipython3
canopy_ppa.crown_profile.projected_crown_area
```

> Those values are the projected *individual* crown areas: the accumulating crown area

Reviewer: ```suggestion
Those values are the cumulative projected crown areas: the accumulating crown area
```

Reviewer: I removed *individual* as this does not seem to refer to individual trees anymore, but to cohort level instead (see comment above).

Author: Not quite - those values are for each individual within each cohort. I've clarified to make this structure clearer.

from the top of the canopy down to the ground. We can take the differences between
vertical layers to give the actual amount of crown area in each layer for each

> individual.

Reviewer: ```suggestion
cohort.
```

Reviewer: This also works at cohort level right (see comments above)?

```{code-cell} ipython3
individual_crown_in_layer = np.diff(
    canopy_ppa.crown_profile.projected_crown_area, prepend=0, axis=0
)
individual_crown_in_layer
```

To bring those values back to the community model: each column represents a cohort, so
we can multiply those values by the number of individuals and then find the row sums
to show the amount of crown area in each layer. As expected, the crown area in each
layer matches the 32 m2 of available space, except for the last layer that is not

> completely filled.

Reviewer: Neat, nice when things add up! 😁
